File: app/src/ultimate/java/com/clockmods/pro/chime/HourlyChimeController.java
package com.clockmods.pro.chime;

import android.os.Handler;
import android.os.Looper;

import com.clockmods.background.BackgroundRepository;

import java.util.Calendar;
import java.util.TimeZone;

public final class HourlyChimeController {
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final RadialChimeView view;
    private final BackgroundRepository repository;
    private long lastChimeAtMillis = Long.MIN_VALUE;
    private final Runnable check = new Runnable() {
        @Override public void run() { checkNow(); handler.postDelayed(this, 250L); }
    };

    public HourlyChimeController(RadialChimeView view, BackgroundRepository repository) {
        this.view = view;
        this.repository = repository;
    }

    public void start() { handler.removeCallbacks(check); handler.post(check); }
    public void stop() { handler.removeCallbacks(check); view.stopChime(); }

    private void checkNow() {
        Calendar now = Calendar.getInstance(resolveTimeZone(repository.getTimeZoneId()));
        now.setTimeInMillis(System.currentTimeMillis());
        long chimeAtMillis = upcomingChimeAtMillis(now, repository.isHourlyChimeEnabled(),
                repository.isHalfHourChimeEnabled());
        if (chimeAtMillis == Long.MIN_VALUE) return;
        Calendar chimeAt = Calendar.getInstance(now.getTimeZone());
        chimeAt.setTimeInMillis(chimeAtMillis);
        if (isQuiet(chimeAt) || chimeAtMillis == lastChimeAtMillis) return;
        lastChimeAtMillis = chimeAtMillis;
        view.startChime(repository, chimeAtMillis);
    }

    static long upcomingChimeAtMillis(Calendar now, boolean hourlyEnabled,
            boolean halfHourEnabled) {
        int minute = now.get(Calendar.MINUTE);
        boolean approachingHour = hourlyEnabled && minute == 59;
        boolean approachingHalfHour = halfHourEnabled && minute == 29;
        if (now.get(Calendar.SECOND) < 58 || (!approachingHour && !approachingHalfHour)) {
            return Long.MIN_VALUE;
        }
        Calendar chimeAt = (Calendar) now.clone();
        chimeAt.add(Calendar.MINUTE, 1);
        chimeAt.set(Calendar.SECOND, 0);
        chimeAt.set(Calendar.MILLISECOND, 0);
        return chimeAt.getTimeInMillis();
    }

    private boolean isQuiet(Calendar now) {
        if (!repository.isHourlyChimeQuietEnabled()) return false;
        int current = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE);
        return isQuietAtMinute(current, repository.getHourlyChimeQuietStart(),
                repository.getHourlyChimeQuietEnd());
    }

    static boolean isQuietAtMinute(int current, int start, int end) {
        current = normalizeMinute(current);
        start = normalizeMinute(start);
        end = normalizeMinute(end);
        if (start == end) return true;
        return start < end ? current >= start && current < end
                : current >= start || current < end;
    }

    static TimeZone resolveTimeZone(String zoneId) {
        if (zoneId == null || zoneId.length() == 0) return TimeZone.getDefault();
        return TimeZone.getTimeZone(zoneId);
    }

    private static int normalizeMinute(int value) {
        int normalized = value % (24 * 60);
        return normalized < 0 ? normalized + 24 * 60 : normalized;
    }
}

File: app/src/ultimate/java/com/clockmods/pro/chime/HourlyChimeController.kt
package com.clockmods.pro.chime

import java.util.Calendar
import java.util.TimeZone

/** Pure scheduling rules used by the Compose chime indicator and tests. */
object HourlyChimeController {
    const val START_BEFORE_MILLIS = 2_000L
    const val DURATION_MILLIS = 5_000L

    /** The cue is anchored to the clock boundary, even if a frame is delivered late. */
    @JvmStatic
    fun progressAtMillis(nowMillis: Long, chimeAtMillis: Long): Float? {
        if (chimeAtMillis == Long.MIN_VALUE) return null
        val elapsed = nowMillis - (chimeAtMillis - START_BEFORE_MILLIS)
        return if (elapsed in 0 until DURATION_MILLIS) elapsed.toFloat() / DURATION_MILLIS else null
    }

    @JvmStatic
    fun upcomingChimeAtMillis(
        now: Calendar,
        hourlyEnabled: Boolean,
        halfHourEnabled: Boolean,
    ): Long {
        val minute = now.get(Calendar.MINUTE)
        val approachingHour = hourlyEnabled && minute == 59
        val approachingHalfHour = halfHourEnabled && minute == 29
        if (now.get(Calendar.SECOND) < 58 || !approachingHour && !approachingHalfHour) {
            return Long.MIN_VALUE
        }
        return (now.clone() as Calendar).apply {
            add(Calendar.MINUTE, 1)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }

    @JvmStatic
    fun isQuietAtMinute(current: Int, start: Int, end: Int): Boolean {
        val normalizedCurrent = normalizeMinute(current)
        val normalizedStart = normalizeMinute(start)
        val normalizedEnd = normalizeMinute(end)
        if (normalizedStart == normalizedEnd) return true
        return if (normalizedStart < normalizedEnd) {
            normalizedCurrent >= normalizedStart && normalizedCurrent < normalizedEnd
        } else {
            normalizedCurrent >= normalizedStart || normalizedCurrent < normalizedEnd
        }
    }

    @JvmStatic
    fun resolveTimeZone(zoneId: String?): TimeZone =
        if (zoneId.isNullOrEmpty()) TimeZone.getDefault() else TimeZone.getTimeZone(zoneId)

    private fun normalizeMinute(value: Int): Int {
        val normalized = value % MINUTES_PER_DAY
        return if (normalized < 0) normalized + MINUTES_PER_DAY else normalized
    }

    private const val MINUTES_PER_DAY = 24 * 60
}
